### 모듈 불러오기

In [1]:
import pandas as pd
from pandas import Series, DataFrame
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns

##### 데이터 불러오기

In [2]:
import os
os.listdir("C:/Users/gun67/전력 프로젝트/2015")

['a. 2015_1_House1_Ch1_Combined_Data_New.csv',
 'b. 2015_2_House1_Ch1_Combined_Data_New.csv',
 'c. 2015_3_House1_Ch1_Combined_Data_New.csv',
 'd. 2015_4_House1_Ch1_Combined_Data_New.csv',
 'e. 2015_5_House1_Ch1_Combined_Data_New.csv',
 'f. 2015_6_House1_Ch1_Combined_Data_New.csv',
 'g. 2015_7_House1_Ch1_Combined_Data_New.csv',
 'h. 2015_8_House1_Ch1_Combined_Data_New.csv',
 'i. 2015_9_House1_Ch1_Combined_Data_New.csv',
 'j. 2015_10_House1_Ch1_Combined_Data_New.csv',
 'k. 2015_11_House1_Ch1_Combined_Data_New.csv',
 'l. 2015_12_House1_Ch1_Combined_Data_New.csv']

In [3]:
cd C:/Users/gun67/전력 프로젝트/2015

C:\Users\gun67\전력 프로젝트\2015


In [4]:
forders = os.listdir("C:/Users/gun67/전력 프로젝트/2015")
df_all_data = pd.DataFrame()

for files in forders :
    df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
    df_all_data = pd.concat([df_all_data, df], axis = 0)

C:\Users\gun67\AppData\Local\Temp\ipykernel_17096\1696432438.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\ipykernel_17096\1696432438.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\ipykernel_17096\1696432438.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Te

In [5]:
df_all_data

Unnamed: 0 Year-Month-Date       Time  Holiday  Watt  Temp('C)  \
0               0      2015-01-01  [ 00:00 ]        1   190       6.0   
1               1      2015-01-01  [ 00:01 ]        1   189       6.0   
2               2      2015-01-01  [ 00:02 ]        1   190       6.0   
3               3      2015-01-01  [ 00:03 ]        1   190       6.0   
4               4      2015-01-01  [ 00:04 ]        1   189       6.0   
...           ...             ...        ...      ...   ...       ...   
44635       44635      2015-12-31  [ 23:55 ]        0  2612       7.0   
44636       44636      2015-12-31  [ 23:56 ]        0  2613       7.0   
44637       44637      2015-12-31  [ 23:57 ]        0  2608       7.0   
44638       44638      2015-12-31  [ 23:58 ]        0  2618       7.0   
44639       44639      2015-12-31  [ 23:59 ]        0   344       7.0   

       Humidity(%)  CPI  
0             76.0  0.3  
1             76.0  0.3  
2             76.0  0.3  
3             76.0  0.3  
4             76.0  0.3  
...            ...  ...  
44635         76.0  0.2  
44636         76.0  0.2  
44637         76.0  0.2  
44638         76.0  0.2  
44639         76.0  0.2  

[525600 rows x 8 columns]

### 불필요한 열 제거

In [6]:
df_all_data.drop(columns = ["Unnamed: 0", "CPI"], inplace = True)
df_all_data

Year-Month-Date       Time  Holiday  Watt  Temp('C)  Humidity(%)
0          2015-01-01  [ 00:00 ]        1   190       6.0         76.0
1          2015-01-01  [ 00:01 ]        1   189       6.0         76.0
2          2015-01-01  [ 00:02 ]        1   190       6.0         76.0
3          2015-01-01  [ 00:03 ]        1   190       6.0         76.0
4          2015-01-01  [ 00:04 ]        1   189       6.0         76.0
...               ...        ...      ...   ...       ...          ...
44635      2015-12-31  [ 23:55 ]        0  2612       7.0         76.0
44636      2015-12-31  [ 23:56 ]        0  2613       7.0         76.0
44637      2015-12-31  [ 23:57 ]        0  2608       7.0         76.0
44638      2015-12-31  [ 23:58 ]        0  2618       7.0         76.0
44639      2015-12-31  [ 23:59 ]        0   344       7.0         76.0

[525600 rows x 6 columns]

### 인덱스 초기화

In [7]:
df_all_data.reset_index(drop = True, inplace = True)
df_all_data

Year-Month-Date       Time  Holiday  Watt  Temp('C)  Humidity(%)
0           2015-01-01  [ 00:00 ]        1   190       6.0         76.0
1           2015-01-01  [ 00:01 ]        1   189       6.0         76.0
2           2015-01-01  [ 00:02 ]        1   190       6.0         76.0
3           2015-01-01  [ 00:03 ]        1   190       6.0         76.0
4           2015-01-01  [ 00:04 ]        1   189       6.0         76.0
...                ...        ...      ...   ...       ...          ...
525595      2015-12-31  [ 23:55 ]        0  2612       7.0         76.0
525596      2015-12-31  [ 23:56 ]        0  2613       7.0         76.0
525597      2015-12-31  [ 23:57 ]        0  2608       7.0         76.0
525598      2015-12-31  [ 23:58 ]        0  2618       7.0         76.0
525599      2015-12-31  [ 23:59 ]        0   344       7.0         76.0

[525600 rows x 6 columns]

### 데이터 타입 확인

In [8]:
print(df_all_data.dtypes)

Year-Month-Date     object
Time                object
Holiday              int64
Watt                object
Temp('C)           float64
Humidity(%)        float64
dtype: object


##### 숫자형 데이터로 표시되야 할 Watt 열에 잘못 기입된 데이터가 있다고 판단  
##### 데이터 타입을 숫자형으로 바꾸기 테스트 진행

In [9]:
df_all_data["Watt"] = pd.to_numeric(df_all_data["Watt"], errors = 'coerce')
print(df_all_data.dtypes)

Year-Month-Date     object
Time                object
Holiday              int64
Watt                 int64
Temp('C)           float64
Humidity(%)        float64
dtype: object


##### 다행히 Watt 열의 데이터 타입이 수치형으로 변경되었음을 알 수 있음

### 결손치 확인

In [10]:
print(df_all_data[df_all_data['Watt'] == 0])
print("-----------------------------------------")
print(df_all_data[df_all_data["Temp('C)"] == "NaN"])
print("-----------------------------------------")
print(df_all_data[(df_all_data["Humidity(%)"] < 0) | (df_all_data["Humidity(%)"] > 100)])

       Year-Month-Date       Time  Holiday  Watt  Temp('C)  Humidity(%)
173489      2015-05-01  [ 11:29 ]        0     0    11.000       54.000
173490      2015-05-01  [ 11:30 ]        0     0    11.000       54.000
173491      2015-05-01  [ 11:31 ]        0     0    11.000       53.867
173492      2015-05-01  [ 11:32 ]        0     0    11.000       53.733
173493      2015-05-01  [ 11:33 ]        0     0    11.000       53.600
...                ...        ...      ...   ...       ...          ...
345357      2015-08-28  [ 19:57 ]        0     0    17.100       67.100
345358      2015-08-28  [ 19:58 ]        0     0    17.067       67.400
345359      2015-08-28  [ 19:59 ]        0     0    17.000       68.000
345360      2015-08-28  [ 20:00 ]        0     0    17.000       68.000
345361      2015-08-28  [ 20:01 ]        0     0    17.000       68.133

[9035 rows x 6 columns]
-----------------------------------------
Empty DataFrame
Columns: [Year-Month-Date, Time, Holiday, Watt, Temp(

##### Watt 열에서는 525600개의 레코드 중 9035개, 약 1.7%의 결손치가 확인 됨  
##### Temp('C) 열과 Humidicy(%) 열은 결손치가 없다고 판단

### 결손치 처리

In [11]:
# 전력값 결손치를 평균값으로 대체
data_watt = pd.DataFrame(df_all_data["Watt"])
print("결손치 대체 전 데이터")
print(data_watt)

# 0이 아닌 값을 가지는 데이터의 평균 계산
Watt_mean = data_watt[data_watt["Watt"] != 0]["Watt"].mean()

# 결손치를 평균값으로 대체
df_all_data.loc[data_watt["Watt"] == 0, "Watt"] = Watt_mean

print("\n결손치 대체 후 데이터")
print(data_watt)

결손치 대체 전 데이터
        Watt
0        190
1        189
2        190
3        190
4        189
...      ...
525595  2612
525596  2613
525597  2608
525598  2618
525599   344

[525600 rows x 1 columns]

결손치 대체 후 데이터
        Watt
0        190
1        189
2        190
3        190
4        189
...      ...
525595  2612
525596  2613
525597  2608
525598  2618
525599   344

[525600 rows x 1 columns]


C:\Users\gun67\AppData\Local\Temp\ipykernel_17096\3442396337.py:10: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '455.4541016135433' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df_all_data.loc[data_watt["Watt"] == 0, "Watt"] = Watt_mean


### 이상치 확인

In [12]:
# Watt

# 통계적 방법으로 이상치 확인
# Z-점수 ; 각 데이터 포인트가 평균에서 몇 표준편차 떨어져 있는지 나타냄, 일반적으로 Z-점수가 3 이상이거나 -3 이하인 경우 이상치로 간주
# 사분위 범위 ; Q1(1사분위수)와 Q3(3사분위수) 사이의 범위를 계산하고, Q1 - 1.5 * IQR 보다 작거나 Q3 + 1.5 * IQR 보다 큰 값을 이상치로 간주

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Watt"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Watt"].quantile(0.25)
Q3 = df_all_data["Watt"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Watt"] < (Q1 - 1.5 * IQR)) | (df_all_data["Watt"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_data_watt = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_data_watt)

Z-scores :  0         0.456561
1         0.458281
2         0.456561
3         0.456561
4         0.458281
            ...   
525595    3.709093
525596    3.710813
525597    3.702214
525598    3.719413
525599    0.191693
Name: Watt, Length: 525600, dtype: float64
IQR :  228.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday    Watt  Temp('C)  Humidity(%)
380         2015-01-01  [ 06:20 ]        1   906.0     8.000       87.000
381         2015-01-01  [ 06:21 ]        1   906.0     8.000       87.000
454         2015-01-01  [ 07:34 ]        1   862.0     8.867       82.667
793         2015-01-01  [ 13:13 ]        1  1046.0     9.000       87.000
2945        2015-01-03  [ 01:05 ]        1   977.0     6.000       66.000
...                ...        ...      ...     ...       ...          ...
525594      2015-12-31  [ 23:54 ]        0  2615.0     7.000       76.000
525595      2015-12-31  [ 23:55 ]        0  2612.0     7.000       76.000
525596      2015-12-31  [ 23:56 

In [13]:
# Temp('C)

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Temp('C)"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Temp('C)"].quantile(0.25)
Q3 = df_all_data["Temp('C)"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Temp('C)"] < (Q1 - 1.5 * IQR)) | (df_all_data["Temp('C)"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_df_Temp = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_df_Temp)

Z-scores :  0         1.265952
1         1.265952
2         1.265952
3         1.265952
4         1.265952
            ...   
525595    1.073706
525596    1.073706
525597    1.073706
525598    1.073706
525599    1.073706
Name: Temp('C), Length: 525600, dtype: float64
IQR :  7.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday        Watt  Temp('C)  Humidity(%)
32086       2015-01-23  [ 06:46 ]        0  192.000000    -1.533       89.733
32087       2015-01-23  [ 06:47 ]        0  195.000000    -1.567       89.967
32088       2015-01-23  [ 06:48 ]        0  192.000000    -1.600       90.200
32089       2015-01-23  [ 06:49 ]        0  192.000000    -1.633       90.433
32090       2015-01-23  [ 06:50 ]        0  229.000000    -1.667       90.667
...                ...        ...      ...         ...       ...          ...
336660      2015-08-22  [ 19:00 ]        1  455.454102    27.000       42.000
336661      2015-08-22  [ 19:01 ]        1  455.454102    26.900       4

In [14]:
# humidity(%)

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Humidity(%)"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Humidity(%)"].quantile(0.25)
Q3 = df_all_data["Humidity(%)"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Humidity(%)"] < (Q1 - 1.5 * IQR)) | (df_all_data["Humidity(%)"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_df_humidity = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_df_humidity)

Z-scores :  0         0.326978
1         0.326978
2         0.326978
3         0.326978
4         0.326978
            ...   
525595    0.326978
525596    0.326978
525597    0.326978
525598    0.326978
525599    0.326978
Name: Humidity(%), Length: 525600, dtype: float64
IQR :  20.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
101586      2015-03-12  [ 13:06 ]        0  371.0    13.200       31.800
101587      2015-03-12  [ 13:07 ]        0  390.0    13.233       31.600
101588      2015-03-12  [ 13:08 ]        0  365.0    13.267       31.400
101589      2015-03-12  [ 13:09 ]        0  365.0    13.300       31.200
101590      2015-03-12  [ 13:10 ]        0  361.0    13.333       31.000
...                ...        ...      ...    ...       ...          ...
317784      2015-08-09  [ 16:24 ]        1  206.0    26.000       31.200
317785      2015-08-09  [ 16:25 ]        1  206.0    26.000       31.333
317786      2015-08-09  [ 16:26 ]  

### 이상치 처리

In [15]:
# Watt

# 이상치 대체 ; 평균, 중앙값 등으로 대체
# 이상치를 중앙값으로 대체
median_value = df_all_data["Watt"].median()
df_all_data.loc[(z_scores >= 3), "Watt"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday    Watt  Temp('C)  Humidity(%)
0           2015-01-01  [ 00:00 ]        1   190.0       6.0         76.0
1           2015-01-01  [ 00:01 ]        1   189.0       6.0         76.0
2           2015-01-01  [ 00:02 ]        1   190.0       6.0         76.0
3           2015-01-01  [ 00:03 ]        1   190.0       6.0         76.0
4           2015-01-01  [ 00:04 ]        1   189.0       6.0         76.0
...                ...        ...      ...     ...       ...          ...
525595      2015-12-31  [ 23:55 ]        0  2612.0       7.0         76.0
525596      2015-12-31  [ 23:56 ]        0  2613.0       7.0         76.0
525597      2015-12-31  [ 23:57 ]        0  2608.0       7.0         76.0
525598      2015-12-31  [ 23:58 ]        0  2618.0       7.0         76.0
525599      2015-12-31  [ 23:59 ]        0   344.0       7.0         76.0

[525600 rows x 6 columns]


In [16]:
# Temp

# 이상치를 중앙값으로 대체
median_value = df_all_data["Temp('C)"].median()
df_all_data.loc[(z_scores >= 3), "Temp('C)"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday    Watt  Temp('C)  Humidity(%)
0           2015-01-01  [ 00:00 ]        1   190.0       6.0         76.0
1           2015-01-01  [ 00:01 ]        1   189.0       6.0         76.0
2           2015-01-01  [ 00:02 ]        1   190.0       6.0         76.0
3           2015-01-01  [ 00:03 ]        1   190.0       6.0         76.0
4           2015-01-01  [ 00:04 ]        1   189.0       6.0         76.0
...                ...        ...      ...     ...       ...          ...
525595      2015-12-31  [ 23:55 ]        0  2612.0       7.0         76.0
525596      2015-12-31  [ 23:56 ]        0  2613.0       7.0         76.0
525597      2015-12-31  [ 23:57 ]        0  2608.0       7.0         76.0
525598      2015-12-31  [ 23:58 ]        0  2618.0       7.0         76.0
525599      2015-12-31  [ 23:59 ]        0   344.0       7.0         76.0

[525600 rows x 6 columns]


In [17]:
# Humidity

# 이상치를 중앙값으로 대체
median_value = df_all_data["Humidity(%)"].median()
df_all_data.loc[(z_scores >= 3), "Humidity(%)"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday    Watt  Temp('C)  Humidity(%)
0           2015-01-01  [ 00:00 ]        1   190.0       6.0         76.0
1           2015-01-01  [ 00:01 ]        1   189.0       6.0         76.0
2           2015-01-01  [ 00:02 ]        1   190.0       6.0         76.0
3           2015-01-01  [ 00:03 ]        1   190.0       6.0         76.0
4           2015-01-01  [ 00:04 ]        1   189.0       6.0         76.0
...                ...        ...      ...     ...       ...          ...
525595      2015-12-31  [ 23:55 ]        0  2612.0       7.0         76.0
525596      2015-12-31  [ 23:56 ]        0  2613.0       7.0         76.0
525597      2015-12-31  [ 23:57 ]        0  2608.0       7.0         76.0
525598      2015-12-31  [ 23:58 ]        0  2618.0       7.0         76.0
525599      2015-12-31  [ 23:59 ]        0   344.0       7.0         76.0

[525600 rows x 6 columns]


### 데이터셋 통합

In [18]:
Watt_N = pd.DataFrame(df_all_data["Watt"], columns = ["Watt"])
Temp_N = pd.DataFrame(df_all_data["Temp('C)"], columns = ["Temp('C)"])
Humidity_N = pd.DataFrame(df_all_data["Humidity(%)"], columns = ["Humidity(%)"])

In [19]:
df_all = pd.concat([df_all_data["Year-Month-Date"], df_all_data["Time"], Watt_N, Temp_N, Humidity_N], axis = 1)
df_all

Year-Month-Date       Time    Watt  Temp('C)  Humidity(%)
0           2015-01-01  [ 00:00 ]   190.0       6.0         76.0
1           2015-01-01  [ 00:01 ]   189.0       6.0         76.0
2           2015-01-01  [ 00:02 ]   190.0       6.0         76.0
3           2015-01-01  [ 00:03 ]   190.0       6.0         76.0
4           2015-01-01  [ 00:04 ]   189.0       6.0         76.0
...                ...        ...     ...       ...          ...
525595      2015-12-31  [ 23:55 ]  2612.0       7.0         76.0
525596      2015-12-31  [ 23:56 ]  2613.0       7.0         76.0
525597      2015-12-31  [ 23:57 ]  2608.0       7.0         76.0
525598      2015-12-31  [ 23:58 ]  2618.0       7.0         76.0
525599      2015-12-31  [ 23:59 ]   344.0       7.0         76.0

[525600 rows x 5 columns]

### 데이터 저장

In [20]:
df_all.to_csv("C:/Users/gun67/전력 프로젝트/preprocessing_data/2015_preprocessing_data", index = False, encoding = "utf-8-sig")

### 저장 데이터 확인

In [21]:
pd.read_csv("C:/Users/gun67/전력 프로젝트/preprocessing_data/2015_preprocessing_data", encoding = "utf-8-sig")

Year-Month-Date       Time    Watt  Temp('C)  Humidity(%)
0           2015-01-01  [ 00:00 ]   190.0       6.0         76.0
1           2015-01-01  [ 00:01 ]   189.0       6.0         76.0
2           2015-01-01  [ 00:02 ]   190.0       6.0         76.0
3           2015-01-01  [ 00:03 ]   190.0       6.0         76.0
4           2015-01-01  [ 00:04 ]   189.0       6.0         76.0
...                ...        ...     ...       ...          ...
525595      2015-12-31  [ 23:55 ]  2612.0       7.0         76.0
525596      2015-12-31  [ 23:56 ]  2613.0       7.0         76.0
525597      2015-12-31  [ 23:57 ]  2608.0       7.0         76.0
525598      2015-12-31  [ 23:58 ]  2618.0       7.0         76.0
525599      2015-12-31  [ 23:59 ]   344.0       7.0         76.0

[525600 rows x 5 columns]